demo/memnet/external_memory.py

```diff
@@ -0,0 +1,119 @@
+# limitations under the License.
+
+from paddle.trainer_config_helpers import *
+
+
+class ExternalMemory(object):
+    def __init__(self, name, mem_slot_size, mem_fea_size, ext_mem_initial, is_test=False, scale=5):
+        self.name = name
+        self.mem_slot_size = mem_slot_size
+        self.mem_fea_size = mem_fea_size
+        self.scale = 5
+        self.external_memory = memory(name=self.name,
+                               	size=mem_fea_size*mem_slot_size,
+                                boot_bias= ParamAttr(initial_std=0.01,
+                                              initial_mean=0.))
+        self.is_test = is_test
+
+    def read(self, read_key):
+        cosine_similarity_read = cos_sim(read_key, self.external_memory, scale=self.scale, size=self.mem_slot_size)
+        norm_cosine_similarity_read = mixed_layer(input=
+                                                  identity_projection(cosine_similarity_read),
+                                                  bias_attr = False,
+                                                  act = SoftmaxActivation(),
+                                                  size = self.mem_slot_size,
+                                                  name='read_weight')
+
+        memory_read = linear_comb_layer(weights=norm_cosine_similarity_read, 
+                                        vectors=self.external_memory,
+                                        size=self.mem_fea_size, name='read_content')
+
+        if self.is_test:
+            print_layer(input=[norm_cosine_similarity_read, memory_read])
+
+        return memory_read
+
+    def write(self, write_key):
+        cosine_similarity_write = cos_sim(write_key, self.external_memory, 
+                                          scale=self.scale, size=self.mem_slot_size)
+        norm_cosine_similarity_write = mixed_layer(input=
+                                                   identity_projection(cosine_similarity_write),
+                                                   bias_attr = False,
+                                                   act = SoftmaxActivation(),
+                                                   size = self.mem_slot_size,
+                                                   name='write_weight')
+        if self.is_test:
+            print_layer(input=[norm_cosine_similarity_write])
+    
+        add_vec = mixed_layer(input = full_matrix_projection(write_key),
+                              bias_attr = None,
+                              act = SoftmaxActivation(),
+                              size = self.mem_fea_size,
+                              name='add_vector')
+
+        erase_vec = self.MakeConstantVector(self.mem_fea_size, 1.0, write_key)
+
+
+        if self.is_test:
+            print_layer(input=[erase_vec])
+            print_layer(input=[add_vec])
+
+        out_prod = out_prod_layer(norm_cosine_similarity_write, erase_vec, name="outer")
```

Comment:
Creating a constant vector erase_vec for this is very ugly. A nicer way to do this is to enhance "repeat" layer to allow repeat in both directions, similar to "repmat" in matlat.